addons/im_livechat/models/mail_channel.py

```diff
@@ -191,3 +194,134 @@ def _email_livechat_transcript(self, email):
             'body_html': mail_body,
         })
         mail.send()
+
+    def _get_channel_history(self):
+        """
+        Converting message body back to plaintext for correct data formatting in HTML field.
+        """
+        return ''.join(
+            '%s: %s<br/>' % (message.author_id.name or self.anonymous_name, html2plaintext(message.body))
+            for message in self.message_ids.sorted('id')
+        )
+
+    # =======================
+    # Chatbot
+    # =======================
+
+    def _fetch_user_input_values(self, inputs_dict):
+        """
+        Look for user's input in the channel's messages based on a dictionary
+        mapping the step_type to the field name of the model it will be used on.
+        """
+        values = {}
+        filtered_message_ids = self.livechat_chatbot_message_ids.filtered(
+            lambda m: m.chatbot_step_id.step_type in inputs_dict.keys()
+        )
+        for message_id in filtered_message_ids:
+            field_name = inputs_dict[message_id.chatbot_step_id.step_type]
+            if not values.get(field_name):
+                values[field_name] = message_id.chatbot_user_raw_answer
+
+        return values
+
+    def _get_user_email_phone(self):
+        """
+        helper method to get user's email and phone
+        """
+        user_input = self._fetch_user_input_values({
+            'question_email': 'email',
+            'question_phone': 'phone',
+        })
+        return user_input.get('email', False), user_input.get('phone', False)
+
+    def _chatbot_post_message(self, chatbot, body, text_to_html=True):
+        if text_to_html:
+            body = plaintext2html(body)
+
+        return self.with_context(mail_create_nosubscribe=True).message_post(
+            author_id=chatbot.operator_partner_id.id,
+            body=body,
+            message_type='comment',
+            subtype_xmlid='mail.mt_comment',
+        )
+
+    def _chatbot_process_step(self, step_id):
+        self.ensure_one()
+        if self.livechat_chatbot_current_step_id and step_id.sequence <= self.livechat_chatbot_current_step_id.sequence:
+            raise ValidationError(_('The new step sequence must be higher than the old one.'))
+
+        # We change the current step to the new step
+        self.livechat_chatbot_current_step_id = step_id
+
+        human_operator = False
+        if step_id.step_type == 'forward_operator':
+            human_operator = self.livechat_channel_id._get_random_operator() if self.livechat_channel_id else False
+            if human_operator:
+                self.with_user(self.env.ref('base.user_root')).add_members(
+                    human_operator.partner_id.ids,
+                    open_chat_window=True,
+                    post_joined_message=False)
+
+                if step_id.message:
```

Comment:
Not sure to understand the whole flow with messages here ^^ .

Reply:
(should be more clear now that it's extracted in sub-methods for specific step types).